src/Type.ml

```diff
@@ -1,23 +1,36 @@
 open Exceptions
+open Core
 
 type t = INT
        | BOOL
        | CHAR
        | STRING
        | LIST
-       [@@deriving sexp]
+       | TVAR of string
+       | ARRAY of (t * t)
+       [@@deriving compare,sexp]
 
-let of_string : string -> t = function
+let of_atomic_string (s:string) : t =
+  match s with 
   | "Int"    -> INT
   | "Bool"   -> BOOL
   | "Char"   -> CHAR
   | "String" -> STRING
   | "List"   -> LIST
-  | s -> raise (Parse_Exn ("Could not parse type `" ^ s ^ "`."))
+  | _ -> raise (Parse_Exn ("Could not parse type `" ^ s ^ "`."))
```

Comment:
Why this change?! We need the `s` because it appears in the exception message.

Reply:
Ah okay, the function argument is named `s` too.

Also, I would actually merge this function with the `of_string` one. See my comment on that function below.